### ISSUE TITLE
Running tests should warn about duplicated rules

### ISSUE THREAD

**Issue by dpelle** [languagetool-org/languagetool#327]
I noticed a duplicated rule in the French grammar file which I removed (a copy/paste mistake).

"mvn clean test" is warning about several ill-constructed rules already, but it would be good if it was also warning about duplicated rule patterns.

One way to do it would be to compute a hash code for rules from everything that defines the rule pattern.  "mvn clean test" can then detect duplicated rules by checking whether 2 or more rules have the same hash code. A hash collision could cause false positive in theory, but the risk of such false positive is so tiny that it's unlikely to happen if the hash function is good.

**Comment by erupothu** [comment]
But even one extra space in the duplicated rule might change the hash function by a lot and we would never figure it out. Right??

**Comment by f-knorr** [comment]
I agree with @erupothu . Comparing rule patterns (by hash,...) does not seem very promising to me.
However, we could check whether incorrect sentences (`<example correction="...">`) trigger more than one match (in rules that are not part of the same rulegroup). If there exists a (partially) duplicated rule, these rules will trigger a match.
Of course, if the error-example contains more than one error, two matches will be triggered as well. In this case, it might be a good idea to modify the error-example.

**Comment by dpelle** [comment]
@erupothu  wrote:

>  But even one extra space in the duplicated rule might change
> the hash function by a lot and we would never figure it out. Right??

What do you mean by extra space? Where would this extra space be?
The hash I had in mind would not be a hash of the xml snippet, but of the parsed xml rule.
So irrelevant spaces such as indentation would not be taken into account.

For duplicated rule detection, only the pattern part of the rule should be taken into
account I think.  So 2 rules with the same pattern are dupe, even if they have
different names, or different examples, etc.

@f-knorr wrote:

> we could check whether incorrect sentences (<example correction="...">) trigger more than one match

That seems unrelated to me. There are plenty of cases where an incorrect sentence
trigger multiple rules and yet the rules are not dupes.

**Comment by f-knorr** [comment]
> @f-knorr wrote:
> 
>     we could check whether incorrect sentences () trigger more than one match
> 
> That seems unrelated to me. There are plenty of cases where an incorrect sentence
> trigger multiple rules and yet the rules are not dupes.

@dpelle : I think there are too many options to write identical rules (eg., `postag="SUB:.*"` vs. `postag="SUB.*"` [either with **:** or without **:**, which would make no difference in a German rule] or `<token regexp="yes">C|[AB]</token>` vs. `<token regexp="yes">[BA]|C</token>`) which would most likely not be covered even by a comparison of hash values of the parsed rule.
On the other hand, checks for overlapping rules would find these duplicated patterns independent from the underlying xml. Of course, there are cases where sentences trigger several matches and it is correct.
However, i just checked multiple matches of German examples (the check is commented out in PatternRuleTest) and i could immediately identify duplications: 
```
WARN: de: 'Du brauchst nicht kommen.' in BRAUCHEN_ZU also matched BRAUCHEN_OHNE_ZU
WARN: de: 'Das funktioniert doch nie im leben!' in IM_LEBEN also matched PRAEP_PLUS_VERB
WARN: de: 'Der Wagen kam ins schleudern.' in INS_SCHLEUDERN also matched PRAEP_PLUS_VERB
```
Hence, option 1 (comparing rules) will probably find not all dupliactions -- especially if an added more general rules also covers an existing special case rule. Option 2 (checking for overlapping matches) will probably find some / many false positives

**Comment by TiagoSantos81** [comment]
> However, i just checked multiple matches of German examples (the check is commented out in PatternRuleTest) and i could immediately identify duplications:

This is a great idea. 
It could be a test on the wikipedia corpus used in the daily regression testing. The matches of each rule would be stored and compared. When two rules have the same match set, a warning message would appear making reference to the said rules. It is not perfect, but would assist greatly in finding duplicated or superseeded rules.

Since this test would be more resource intensive than the regular regression testing, it could be automated to run only once a week, or once a month

Later it could be extended to advise on rules that that aproximated match sets, so that rule designer become aware of rules merging opportunities.

**Comment by dpelle** [comment]
@f-knorr wrote:

> On the other hand, checks for overlapping rules would find these duplicated patterns
> independent from the underlying xml.

While this may be useful, I think it's going to be quite noisy. If there are many false positives, developers will tend to ignore them.

My suggestion instead will have 0 false positive, i.e. any reported duplicated rule is expected to be a mistake in xml rules.  So it can be turned on by default when running tests.  I understand that slight changes in the way rules are written may cause to not detect all dupes though (false negatives).

Your suggestion would not be suitable when running tests in my opinion if has many false positive.  But it's still useful if such a check once in a while, and manually inspect potential duplicates.

**Comment by f-knorr** [comment]
> While this may be useful, I think it's going to be quite noisy. If there are many false positives, developers will tend to ignore them.

I totally agree with you. The suggested check will definitively be noisy.

> My suggestion instead will have 0 false positive, i.e. any reported duplicated rule is expected to be a mistake in xml rules. So it can be turned on by default when running tests. I understand that slight changes in the way rules are written may cause to not detect all dupes though (false negatives).

Again: I agree. Yet, I doubt that it makes sense to invest time in a new check that is so sensible to input data, whereas the first test already exists. To cope with the noisy output of the first approach, we might include this output to the daily regression tests and show only diff to the previous run (as we already do it). Thereby, one could easily identify overlapping errors resulting from recently introduced changes.

**Comment by Discostu36** [comment]
> To cope with the noisy output of the first approach, we might include this output to the daily regression tests and show only diff to the previous run (as we already do it). Thereby, one could easily identify overlapping errors resulting from recently introduced changes.

I think more than a year later this is still a good idea. couldn't it be implemented? (Or is it already and I haven't noticed?)
